File: web/lib/decode.js
// Decodes an arming payment back out of its own bytes, into English.
//
// This exists to be *distrustful of the encoder next to it*. The page could simply restate the form
// the user filled in — and that would look identical while proving nothing, because a bug in the
// encoder would be echoed back word for word. So nothing here reads the form. Everything below is
// recovered from the 42-byte memo and the user-operation pre-image, the same two artefacts an
// executor sees, and the commitment is re-derived rather than trusted.
//
// It mirrors `arming/bin/decode.dart`, which does the same job at the command line and exits
// non-zero when a memo does not match its pre-image.

import { keccak256 } from './keccak.js';
import { toHex } from './arming.js';

const VERBS = { 0: 'sell it on a market', 1: 'deposit it into the yield vault', 2: 'withdraw from the yield vault' };
const TRIGGERS = { 0: 'PRICE_BELOW', 1: 'PRICE_ABOVE', 2: 'SCHEDULE', 3: 'PRIVATE' };

const word = (b, off) => {
  let v = 0n;
  for (let i = 0; i < 32; i++) v = (v << 8n) | BigInt(b[off + i]);
  return v;
};
const addressAt = (b, off) => '0x' + toHex(b.slice(off + 12, off + 32));
const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));

const fmtUnits = (v, decimals) => {
  const d = BigInt(10) ** BigInt(decimals);
  const whole = v / d;
  const frac = (v % d).toString().padStart(decimals, '0').replace(/0+$/, '');
  return frac ? `${whole}.${frac}` : `${whole}`;
};

const fmtDuration = (s) => {
  const n = Number(s);
  if (n % 86400 === 0) return n === 86400 ? 'day' : `${n / 86400} days`;
  if (n % 3600 === 0) return n === 3600 ? 'hour' : `${n / 3600} hours`;
  if (n % 60 === 0) return n === 60 ? 'minute' : `${n / 60} minutes`;
  return `${n} seconds`;
};

/**
 * @param memo    the 42 bytes that will travel in the XRPL payment
 * @param userOp  abi.encode(PackedUserOperation) — what the memo commits to
 */
export function decodeArmingPayment(memo, userOp, ctx = {}) {
  const problems = [];

  // ---- the memo, on its own terms -------------------------------------------------------------
  if (memo.length !== 42) problems.push(`the memo is ${memo.length} bytes, and must be exactly 42`);
  if (memo[0] !== 0xfe) problems.push(`the memo opcode is 0x${memo[0].toString(16)}, expected 0xFE`);

  let fee = 0n;
  for (let i = 2; i < 10; i++) fee = (fee << 8n) | BigInt(memo[i]);
  if (fee === 0n) {
    problems.push('the executor fee is zero, so no executor will ever run this payment');
  }

  // ---- does the memo actually commit to this batch? -------------------------------------------
  const commitment = keccak256(userOp);
  const carried = memo.slice(10, 42);
  const matches = carried.length === 32 && commitment.every((b, i) => b === carried[i]);
  if (!matches) {
    problems.push('the memo does not commit to this user operation — the payment would do something else');
  }

  // ---- the batch itself ------------------------------------------------------------------------
  // Layout is fixed by the encoder: after the leading offset word, the head is nine words, and the
  // callData tail begins at the offset the fifth word points at.
  const calls = [];
  try {
    // Head, after the leading offset-to-struct word:
    //   0 sender  1 nonce  2 initCode*  3 callData*  4 accountGasLimits
    //   5 preVerificationGas  6 gasFees  7 paymasterAndData*  8 signature*
    // Slot 4 is a zero-filled value, not a pointer — reading callData from there yields offset 0
    // and the parse walks off the end.
    const base = 32;
    const sender = addressAt(userOp, base);
    const nonce = word(userOp, base + 32);
    const callDataOff = base + Number(word(userOp, base + 32 * 3));
    const callDataLen = Number(word(userOp, callDataOff));
    const cd = userOp.slice(callDataOff + 32, callDataOff + 32 + callDataLen);

    const arrayBase = 4 + 32 + 32;                     // selector, offset-to-array, array length
    const count = Number(word(cd, 4 + 32));
    for (let i = 0; i < count; i++) {
      const elemOff = arrayBase + Number(word(cd, arrayBase + i * 32));
      const target = addressAt(cd, elemOff);
      const dataOff = elemOff + Number(word(cd, elemOff + 64));
      const dataLen = Number(word(cd, dataOff));
      calls.push({ target, data: cd.slice(dataOff + 32, dataOff + 32 + dataLen) });
    }
    ctx = { ...ctx, sender, nonce };
  } catch (e) {
    // Say what went wrong. A bare "could not be decoded" is the kind of message that turns a
    // five-minute diagnosis into an hour of guessing.
    problems.push(
      `the user operation could not be decoded, so what it would do is unknown (${e.message})`,
    );
  }

  // ---- render ----------------------------------------------------------------------------------
  let html = '';
  if (problems.length) {
    html += `<div class="banner bad"><strong>This payment is not safe to send.</strong><ul>`
      + problems.map((p) => `<li>${esc(p)}</li>`).join('') + '</ul></div>';
  }

  html += `<p class="note" style="margin-top:0">Signing this payment would do
    ${calls.length} thing${calls.length === 1 ? '' : 's'}, in order:</p><ol class="plain">`;

  for (const call of calls) {
    const sel = toHex(call.data.slice(0, 4));
    if (sel === '095ea7b3') {
      const spender = addressAt(call.data, 4);
      const amount = word(call.data, 36);
      // Name it only when it really is the contract we expect. An unexpected spender must read as
      // unexpected rather than being dressed up in a friendly label.
      const known = ctx.trimmy && spender.toLowerCase() === ctx.trimmy.toLowerCase();
      const who = known
        ? `<strong>Trimmy</strong> <span class="mono">(${esc(spender)})</span>`
        : `<strong>an unrecognised contract</strong> <span class="mono">${esc(spender)}</span>`;
      html += `<li><strong>Allow spending</strong> — let ${who} move up to
        <strong>${fmtUnits(amount, 6)} XRP</strong> of yours.
        <div class="note">This is an allowance, not a transfer. It is capped at exactly this
        amount and nothing can spend more than it.${
          known ? '' : ' <strong>This is not the Trimmy contract. Do not send this.</strong>'
        }</div></li>`;
    } else if (sel === 'cc0c55f4') {
      html += `<li><strong>Create the rule</strong>${renderRule(call.data)}</li>`;
    } else {
      html += `<li><strong>Unrecognised call</strong> to
        <span class="mono">${esc(call.target)}</span> — selector
        <span class="mono">0x${esc(sel)}</span>.
        <div class="note">This page cannot tell you what it does. Do not send it.</div></li>`;
    }
  }
  html += '</ol>';

  html += `<dl class="kv" style="margin-top:1.2rem">
    <dt>Executor fee</dt><dd>${fmtUnits(fee, 6)} XRP</dd>
    <dt>Commitment</dt><dd>0x${toHex(commitment)}</dd>
    <dt>Memo matches</dt><dd>${matches ? 'yes — re-derived from the bytes above' : 'NO'}</dd>
  </dl>`;

  return { html, problems, matches, commitment };
}

function renderRule(data) {
  const at = (i) => word(data, 4 + i * 32);
  const buyTokenId = Number(at(1));
  const verb = Number(at(2));
  const trigger = Number(at(4));
  const total = at(5), part = at(6), minOut = at(7), triggerValue = at(8), expiry = at(9);
  const slippage = Number(at(10)), keeperFee = at(12);

  // The contract's own arithmetic, not an approximation of it: Trimmy computes
  // `maxExecutions = ceilDiv(totalSellAmount, partSellAmount)` and refuses to execute once
  // `spent >= totalSellAmount`. Dividing without the ceiling under-reports a rule whose total is
  // not a whole multiple of its part.
  const runs = part > 0n ? (total + part - 1n) / part : 1n;
  const once = runs === 1n;

  let when;
  if (trigger === 2) {
    // A rule is eligible the moment it is armed (`nextEligibleAt = block.timestamp`), so the
    // interval governs the gap between runs, never the wait for the first one. Saying "every hour"
    // about a rule that can only run once is the exact misreading this decoder exists to prevent.
    when = once
      ? 'once, as soon as somebody runs it for you'
      : `${runs} times — the first straight away, then every ${fmtDuration(triggerValue)}`;
  } else if (trigger === 0) {
    when = `when 1 XRP falls to ${fmtUnits(triggerValue, 18)} FLR or below`;
  } else if (trigger === 1) {
    when = `when 1 XRP rises to ${fmtUnits(triggerValue, 18)} FLR or above`;
  } else if (trigger === 3) {
    when = 'at a secret price held only inside a secure enclave';
  } else {
    when = `trigger #${trigger}`;
  }

  const expiryDate = new Date(Number(expiry) * 1000).toISOString().replace('T', ' ').slice(0, 16);

  let out = `<div style="margin-top:.4rem">`;
  out += `<div>Take <strong>${fmtUnits(part, 6)} XRP</strong> and
    <strong>${esc(VERBS[verb] ?? `verb #${verb}`)}</strong>, <strong>${esc(when)}</strong>.</div>`;
  // State the count in every case. Gating this on `runs > 1` is what let a one-shot rule be
  // described as a recurring one with nothing to correct the impression.
  out += once
    ? `<div class="note">This happens <strong>once</strong>, and then the rule is finished.
       ${trigger === 2 ? 'The interval never applies, because there is no second run.' : ''}</div>`
    : `<div class="note">Up to <strong>${runs}</strong> runs of ${fmtUnits(part, 6)} XRP —
       <strong>${fmtUnits(total, 6)} XRP</strong> in all, if every one of them fires.</div>`;
  out += `<div class="note">Expires ${esc(expiryDate)} UTC. Anything unspent stays yours.</div>`;

  if (trigger === 3) {
    out += `<div class="note"><strong>Nothing on the public ledger reveals your price.</strong>
      The chain stores only a commitment to it; the number itself is delivered to the enclave
      encrypted, and it never leaves.</div>`;
  }
  if (verb === 0) {
    out += `<div class="note">Each sale must land within ${slippage / 100}% of the oracle price,
      or it does not happen at all.</div>`;
  }
  if (verb === 1 && minOut > 0n) {
    out += `<div class="note">Must return at least ${fmtUnits(minOut, 6)} vault shares.</div>`;
  }
  out += `<div class="note">Whoever runs it for you earns ${fmtUnits(keeperFee, 6)} XRP per run,
    paid out of the proceeds. You never pay gas.</div>`;
  out += '</div>';
  return out;
}

File: arming/bin/decode.dart
// Renders what an arming payment will actually do, in plain terms.
//
// ## Why this exists as a separate program
//
// The `0xFE` memo a user signs in Xaman is 42 bytes, 32 of which are a hash. It is unreadable. The
// batch it commits to is delivered to the executor off-chain, so nothing the user sees describes
// where their money is about to go.
//
// A decoder built into the same front end that produced the payment does not fix that: a
// compromised front end takes its own decoder with it, and the user is told a comforting story
// about a hostile batch. So this is a **standalone program with no network access**. It reads a
// pre-image, re-derives the commitment, and describes the calls. And it will do that for a
// payment produced by anyone, including one produced by an attacker.
//
// On stage this decodes a batch we did not generate. That is the point.
//
// Usage:
//   dart run bin/decode.dart --preimage 0x... [--memo FE00...]
//   dart run bin/decode.dart --file out/userop-preimage.hex --memo-file out/memo.hex

import 'dart:io';
import 'dart:typed_data';

import 'package:plimsoll_core/plimsoll_core.dart' as p;

const _fxrp = '0x0b6a3645c240605887a5532109323a3e12273dc7';

const _verbs = ['SWAP', 'DEPOSIT_VAULT', 'EXIT_VAULT'];
const _triggers = ['PRICE_BELOW', 'PRICE_ABOVE', 'SCHEDULE'];

String _hex(Uint8List b) =>
    b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

Uint8List _parseHex(String s) {
  final t = s.trim().replaceFirst(RegExp('^0x', caseSensitive: false), '');
  if (t.length.isOdd) throw ArgumentError('odd-length hex');
  return Uint8List.fromList([
    for (var i = 0; i < t.length; i += 2) int.parse(t.substring(i, i + 2), radix: 16),
  ]);
}

BigInt _word(Uint8List d, int i) {
  var v = BigInt.zero;
  for (var k = i * 32; k < (i + 1) * 32; k++) {
    v = (v << 8) | BigInt.from(d[k]);
  }
  return v;
}

/// 6-decimal amounts, rendered as a decimal string. No floating point anywhere, rule 8.
String _fxrpAmount(BigInt uba) {
  final s = uba.toString().padLeft(7, '0');
  final whole = s.substring(0, s.length - 6);
  final frac = s.substring(s.length - 6).replaceFirst(RegExp(r'0+$'), '');
  return frac.isEmpty ? whole : '$whole.$frac';
}

void _describeApprove(Uint8List data) {
  final spender = '0x${_hex(Uint8List.sublistView(data, 16, 36))}';
  final amount = _word(Uint8List.sublistView(data, 4), 1);
  stdout.writeln('  1. APPROVE');
  stdout.writeln('     lets  $spender');
  stdout.writeln('     spend ${_fxrpAmount(amount)} FXRP of yours');
  stdout.writeln('     note  this is an allowance, not a transfer. It is capped at this exact');
  stdout.writeln('           amount, and nothing may spend more than it.');
}

void _describeArm(Uint8List data) {
  final b = Uint8List.sublistView(data, 4);
  int w(int i) => _word(b, i).toInt();
  BigInt bw(int i) => _word(b, i);

  final verb = w(2);
  final trigger = w(4);
  stdout.writeln('  2. ARM A RULE');
  stdout.writeln('     action     ${verb < _verbs.length ? _verbs[verb] : 'UNKNOWN($verb)'}');
  stdout.writeln('     trigger    ${trigger < _triggers.length ? _triggers[trigger] : 'UNKNOWN($trigger)'}');
  stdout.writeln('     sell up to ${_fxrpAmount(bw(5))} (token #${w(0)})');
  stdout.writeln('     per run    ${_fxrpAmount(bw(6))}');
  stdout.writeln('     min out    ${_fxrpAmount(bw(7))}');

  // Trimmy derives the run count itself as ceilDiv(totalSellAmount, partSellAmount) and refuses to
  // execute once spent >= total. State it, rather than leaving a reader to divide two numbers
  // several lines apart: a rule with total == part fires exactly ONCE, and printing only
  // "every 60 seconds" against it describes a recurring rule that does not exist.
  final total = bw(5);
  final perRun = bw(6);
  final runs = perRun > BigInt.zero
      ? (total + perRun - BigInt.one) ~/ perRun
      : BigInt.one;
  stdout.writeln('     runs       $runs (ceilDiv of the two amounts above)');

  if (trigger == 2) {
    if (runs == BigInt.one) {
      stdout.writeln('     every      n/a, one run only, so the ${bw(8)}s interval never applies');
    } else {
      stdout.writeln('     every      ${bw(8)} seconds, after an immediate first run');
    }
  } else {
    stdout.writeln('     threshold  ${bw(8)} (buy-token units per 1 whole sell token)');
  }
  final expiry = bw(9).toInt();
  stdout.writeln('     expires    ${DateTime.fromMillisecondsSinceEpoch(expiry * 1000, isUtc: true)}');
  // O-17: saying "slippage 50 bips max" alone is misleading, because fees come out of the
  // proceeds afterwards. The floor bounds what the VENUE may take; the fees are separate,
  // disclosed deductions. Both are stated, and so is the worst case the user actually receives.
  final part = bw(6);
  final slip = w(10);
  final keeperFee = bw(12);
  final protoBips = w(11);
  stdout.writeln('     venue floor  at most $slip bips below the oracle price');
  stdout.writeln('     keeper fee   ${_fxrpAmount(keeperFee)} per run '
      '(${_fxrpAmount(bw(13))} lifetime cap)');
  stdout.writeln('     protocol fee $protoBips bips of proceeds');
  stdout.writeln('     -> per run you sell ${_fxrpAmount(part)} and receive the proceeds');
  stdout.writeln('        MINUS those fees. The floor above bounds the venue, not the fees.');
}

void main(List<String> args) {
  // NOTE: `int main()` does NOT set the process exit code in Dart. Every refusal below calls
  // exit() explicitly, so a shell guard or CI step can actually act on it.
  String? arg(String n) {
    final i = args.indexOf('--$n');
    return i >= 0 && i + 1 < args.length ? args[i + 1] : null;
  }

  var preimageHex = arg('preimage');
  final file = arg('file');
  if (preimageHex == null && file != null) preimageHex = File(file).readAsStringSync();
  if (preimageHex == null) {
    stderr.writeln('need --preimage <hex> or --file <path>');
    exit(2);
  }

  final preimage = _parseHex(preimageHex);
  final commitment = p.keccak256(preimage);

  stdout.writeln('Trimmy arming payment, independent decode');
  stdout.writeln('  (no network, no trust in whoever produced this payment)\n');
  stdout.writeln('  user operation : ${preimage.length} bytes');
  stdout.writeln('  commitment     : 0x${_hex(commitment)}');

  // If a memo was supplied, the commitment must match it. This is the check that catches a payment
  // whose visible memo does not correspond to the batch being described.
  final memoHex = arg('memo') ??
      (arg('memo-file') != null ? File(arg('memo-file')!).readAsStringSync() : null);
  if (memoHex != null) {
    final memo = _parseHex(memoHex);
    if (memo.length != 42) {
      stdout.writeln('\n  MEMO REJECTED: ${memo.length} bytes, must be 42');
      exit(1);
    }
    if (memo[0] != 0xFE) {
      stdout.writeln('\n  MEMO REJECTED: opcode 0x${memo[0].toRadixString(16)}, expected 0xFE');
      exit(1);
    }
    final carried = Uint8List.sublistView(memo, 10, 42);
    for (var i = 0; i < 32; i++) {
      if (carried[i] != commitment[i]) {
        stdout.writeln('\n  ****  MEMO DOES NOT MATCH THIS USER OPERATION  ****');
        stdout.writeln('  memo commits to : 0x${_hex(carried)}');
        stdout.writeln('  this batch is   : 0x${_hex(commitment)}');
        stdout.writeln('  DO NOT SIGN. The payment would run a batch other than the one below.');
        exit(1);
      }
    }
    stdout.writeln('  memo           : matches this batch');
  }

  // Decode with Plimsoll, which knows the measured wire format.
  final op = p.PackedUserOperation.decode(preimage);
  switch (op) {
    case p.UserOpUndecodable(:final reason):
      stdout.writeln('\n  UNDECODABLE: $reason');
      exit(1);
    case p.UserOpDecoded(:final op):
      stdout.writeln('  sender         : ${op.sender}');
      stdout.writeln('  nonce          : ${op.nonce}');

      final batch = op.decodeCalls();
      switch (batch) {
        case p.CallBatchUndecodable(:final reason):
          stdout.writeln('\n  BATCH UNDECODABLE: $reason');
          exit(1);
        case p.CallBatchDecoded(:final calls):
          stdout.writeln('\n  This payment will do ${calls.length} thing(s):\n');
          for (final c in calls) {
            final data = c.data;
            if (data.length >= 4) {
              final sel = _hex(Uint8List.sublistView(data, 0, 4));
              if (sel == '095ea7b3') {
                _describeApprove(data);
                stdout.writeln('');
                continue;
              }
              if (sel == 'cc0c55f4') {
                _describeArm(data);
                stdout.writeln('');
                continue;
              }
              stdout.writeln('  ?. UNRECOGNISED CALL to ${c.target}');
              stdout.writeln('     selector 0x$sel. This decoder does not know it.');
              stdout.writeln('     Treat that as a reason NOT to sign, not as a detail.');
              stdout.writeln('');
            }
          }
          final touchesFxrp =
              calls.any((c) => c.target.toLowerCase() == _fxrp);
          if (!touchesFxrp) {
            stdout.writeln('  NOTE: no call touches FXRP. Check that is what you meant.');
          }
      }
  }
  exit(0);
}
